### PR TITLE
Root-finding algorithms (bisection)

### PR DESCRIPTION
**Pull request by dhamalakamal**
# Title
- Root-finding algorithms (bisection)
  - wrapper for scipy implementation 
pure python implementation

## Linting
  - Does your code lint? If not are the errors within acceptable errors?
    Yes

## Contents
  - The file contains the code to find roots using bisection method where implementation of wrapper for scipy  has been done and also pure python implementation python has been carried out

## Labels and Milestones
  - ALL pull requests must have an attached label and Milestone before a reviewer can be assigned.

## Reviewer
  - @HeshamElsaman @ojha-aditya

### CODE REVIEW

**Comment by zbpetersbuf** [comment]
I can be your second reviewer, but some one else needs to review before I do, also your going to need to resolve merge conflicts

**Comment by ojha-aditya** [comment]
I can review this. Should I go ahead @zbpetersbuf ?

File: calculus.py [hunk 1]
```diff
@@ -1,8 +1,105 @@
 """
-calculus.py
+calculus.py module
+
+This module provides implementations of the bisection root-finding algorithm 
+in three ways: using a SciPy wrapper, a pure Python implementation, 
+and a C/C++ implementation via ctypes.
```
Comment (ojha-aditya) [inline calculus.py:6]: The module doctoring could be modified to include something more general as this module would also contain other algorithms.

File: calculus.py [hunk 2]
```diff
@@ -1,8 +1,105 @@
...
+    Parameters:
+        func (callable): The function for which to find the root.
+        a (float): The start of the interval.
+        b (float): The end of the interval.
+        tol (float, optional): The tolerance level for convergence. Defaults to 1e-6.
+        max_iter (int, optional): Maximum number of iterations. Defaults to 1000.
+
+    Returns:
+        float: The approximate root of the function.
+
```
Comment (ojha-aditya) [inline calculus.py:35]: I think it was discussed in class once by @laserlab that docstrings should not contain data types in general. Maybe the variable description is enough in itself.

**Review by ojha-aditya — CHANGES_REQUESTED** [review]
The docstrings can be modified as per the comments and if possible instead of raising ValueErrors, you can try resolving those errors by either changing the function value or value of x at that point or some other means. this was what @laserlab had suggested yesterday.

After this and resolving merge conflicts, it is good to merge.

**Comment by zbpetersbuf** [comment]
ya go ahead with your review as you have done, its just needs my 'review' after yours to better control who gets to merge pull requests

File: calculus.py (hunk 2 above)
Comment (dhamalakamal) [reply calculus.py:312]: updated. thanks!

**Comment by zbpetersbuf** [comment]
@ojha-aditya you must approve the pull request before I can

**Review by ojha-aditya — APPROVED** [review]
Just a suggestion, since scipy.optimise is already imported, that can be used directly instead of importing a specific module from it I believe and if possible raising value error can be replaced with solving that error within the code itself. 

Nevertheless, these are just improvements in functionality. The code can be merged I guess as it is needed for next part of the HW.

**Comment by ojha-aditya** [comment]
Although, @dhamalakamal the branch needs to be updated before @zbpetersbuf can merge the PR.

**Comment by zbpetersbuf** [comment]
ill merge as soon as I see this pull request updated

**Comment by ojha-aditya** [comment]
@dhamalakamal you can update the branch so that it can be merged I believe.

**Comment by ojha-aditya** [comment]
> ill merge as soon as I see this pull request updated

It is now pdated I believe.

**Review by zbpetersbuf — APPROVED** [review]
approve

**Comment by ojha-aditya** [comment]
Okay now @dhamalakamal you would also need to resolve conflicts before @zbpetersbuf can merge this as PR #29 was just mrged it seems.

**Comment by dhamalakamal** [comment]
> Okay now @dhamalakamal you would also need to resolve conflicts before @zbpetersbuf can merge this as PR #29 was just merged it seems.

Resolved conflict... It is ready to merge. Please merge it berofe merging any other PR @zbpetersbuf

**Pull request merged**